apps/emqx_retainer/src/emqx_retainer.erl

```diff
@@ -205,7 +205,7 @@ store_retained(Msg = #message{topic = Topic, payload = Payload}, Env) ->
                                                msg = Msg,
                                                expiry_time = get_expiry_time(Msg, Env)});
         {true, false} ->
-            case mnesia:dirty_read(?TAB, Topic) of
+            case mnesia:read(?TAB, Topic) of
```

Comment:
The `mnesia:read/2` should be warrpped into `mnesia:transaction/1`. Otherwise, it will be crashed here?

So, we can't change the `mnesia:dirty_read/2` to `mnesia:read/2` directly

Reply:
thank you, fixed to 

```
{true, false} ->
            {atomic, _} = mnesia:transaction(
                fun() ->
                    case mnesia:read(?TAB, Topic) of
                        [_] ->
                            mnesia:write(?TAB, #retained{topic = topic2tokens(Topic),
                                                         msg = Msg,
                                                         expiry_time = get_expiry_time(Msg, Env)}, write);
                        [] ->
                            ?LOG(error, "Cannot retain message(topic=~s) for table is full!", [Topic])
                    end
                end),
            ok;
```